feat(render): support pseudo-transparency

bennofs opened the pull request:
This adds pseudo-transparency for the background of the bar and the
background of the systray.

Here's how it lools (systray is on the right edge, after the power symbol):

![pic](https://cloud.githubusercontent.com/assets/3192959/26372954/078f84ca-4000-11e7-8c47-49dd34321e6c.png)

I've submitted the PR now to get some feedback on the general implementation, there's still a few small cleanup tasks to do:

- [ ] Cleanup tray changes
- [ ] Check behaviour with compositing enabled.

bennofs commented:
@jaagr any comments?

jaagr commented:
@beenofs sorry for my extreme delay. I have a few comments, will open up a review as soon as I can.

bennofs commented:
@jaagr no problem, take your time. Just wanted to make sure this didn't get lost, as I often forget about PRs myself when I only take a short look at them and then github notification is gone ☺

lmoesch commented:
This feature is currently not compatible with a bottom aligned bar.

bennofs commented:
@lmoesch oh? what happens in that case? i had assumed that `m_rect` would be the correct rect for a bottom alligned bar as well

lmoesch commented:
The background area is still copied from the top of the background, not the bottom. I tried to modify the values of "copy_area_checked" within the background_manager wich did the trick. So basically we need to get the corner point information to that function.

bennofs commented:
@lmoesch background_manager stores a `m_rect` that specifies the target location from which the background should be copied. It is passed to it in via the `activate` method, so we just need to make sure we pass it the right values there.

bennofs commented:
@lmoesch should be fixed with last commit.

lmoesch commented:
Splendid, that worked like a charm! Thanks for the quick fix 👍

lmoesch commented:
While fiddeling arround with some parameters I came across other bugs. 

First thing is, that the offset of a border is not taken into account while calculating the background snippet. Unfortunately I have just few to no knowledge about X-Server classes and behaviour, but I assume that a border is also drawn by a rectangle which gets overdrawn by the content. So to fix this one I guess it is sufficient to replace the bar rectangle with the border rectangle.

Other thing is noticed is, that offset parameters are not working anymore [neither global/wm offset nor bar offset]. The background is drawn here from the right position, but the bar stays at the same place. 
Not sure if this is an issue with this commit, or some general bug (or misconfiguration of my setup).

Last thing is a trickey one. When using multiple bars ontop of each other, background location also fails due to the offset to each other. (e.g. if you have two bars on top, than both will have the background of the most upper one)

bennofs commented:
@lmoesch 

Borders work correctly now: (this is a semi-transparent border at top)

![polybar](https://user-images.githubusercontent.com/3192959/29924918-e7526a1c-8e5e-11e7-9adb-0a7f6f6c8119.png)

Have to look into the "multipe bars stacked" issue.

bennofs commented:
Btw, I am not sure anymore what I originally meant by "Cleanup try changes"

bennofs commented:
Ok, fixed the second issue as well

![polybar](https://user-images.githubusercontent.com/3192959/29933953-bda39ec8-8e79-11e7-922c-3b24a179f3e7.png)

Hope I haved missed any configuration :)

bennofs commented:
Rebased against latest master

NBonaparte commented on src/components/parser.cpp:
```diff
@@ -77,7 +77,7 @@ void parser::codeblock(string&& data, const bar_settings& bar) {
     switch (tag) {
       case 'B':
-        m_sig.emit(change_background{parse_color(value, bar.background)});
+        m_sig.emit(change_background{parse_color(value, 0x0)});
```
Why is this set to `0x0`?

bennofs replied:
This is because 0 specifes "fully transparent": the bar background is now handled in the renderer by compositing everything on top of a surface filled with the background using cairo. So to use "background color" we can simply use "transparent".

bennofs commented:
I have changed to using brace initialization. The tray works fine for me, but alpha support in the tray is dependent on the tray icon having alpha channel I think. What program have you tried, maybe it is specific to that program that it does not work (does it work if using a compositor)? I am using `udiskie` which has a tray icon and that it works there for me. 

Yes we need a detection method. I will look into that. But with the changes, it should be pretty easy to turn of the pseudo-transparency if there is a compositor: right now, we fill the surface with the desktop background as a first step and then paint everything on top of that, using cairo's support for transparency. If we have a compositor, we can simply drop the first step and paint the window fully transparent, everything else should keep working.

NBonaparte commented:
I've tried `nm-applet` and `yakyak` (an Electron app) and both result in an opaque background, with and without compton.

bennofs commented:
@NBonaparte hmm, if I use the example config and change `background` to `#BB222222` everything is semi-transparent for me. Were you using `tray-background`?

bennofs commented:
It works for me even when using `tray-background` (if `tray-background` is set, it takes precendence over `background` and is used instead of it) or `tray-transparent` (which will cause the tray to become 100% transparent)

NBonaparte commented:
I tried with and without `tray-background`, and it just gives me an opaque color. With #BB222222 I get the color #191919.

bennofs commented:
What window manager, desktop, distribution are you using? You're using the default config except for changing the background color, right? I can't reproduce this here.

NBonaparte commented:
Using bspwm, Arch Linux, default config with `background = #BB222222` results in this:
![2017-12-21-150845_53x27_scrot](https://user-images.githubusercontent.com/12215509/34278625-e5726594-e660-11e7-8a09-f6f10dc3c470.png)

bennofs commented:
Hmm, are you setting a background image? I use `feh --bg-fill /path/to/desktop/background.jpg`

NBonaparte commented:
Yeah, using `nitrogen`. The rest of the bar looks like 
![2017-12-21-151734_1920x35_scrot](https://user-images.githubusercontent.com/12215509/34278830-2333b422-e662-11e7-85b4-01f99eb7c412.png)

bennofs commented:
That is very weird. I just launched bspwm, then used `nitrogen --set-scaled --random /some/dir` and after that, launched polybar from terminal with background set to `#22FFFFFF` (used this color since the bg image was very dark so the effect is better seen with a white semi-transparent background) and I got this:

![2017-12-22-002521_1919x41_scrot](https://user-images.githubusercontent.com/3192959/34279069-ff4b6616-e6ae-11e7-8dcd-0a8b5d128748.png)

Everything is neatly transparent...

NBonaparte commented:
@patrick96 Could you test this out and see if the tray pseudo-transparency works?

patrick96 commented:
I have two icons on the tray: dropbox and f.lux.
The tray background itself is fully transparent and the dropbox icon takes the bar transparent background (when compton is running) and the f.lux icon uses a fully transparent background

**With compton:**
![compton](https://user-images.githubusercontent.com/2452038/34446516-eb8da566-ecdb-11e7-85ac-c8e29cc32b5d.png)

**Without compton:**
![no_compton](https://user-images.githubusercontent.com/2452038/34446517-ebc3f76a-ecdb-11e7-9a8a-4f7c8998f8fb.png)

This is the config I used:
```dosini
[colors]
background = #44222222
foreground = #dfdfdf

[bar/example]
monitor = ${env:MONITOR:HDMI-0}
width = 100%
height = 27
radius = 6.0

background = ${colors.background}
foreground = ${colors.foreground}

border-size = 4
border-color = #00000000

padding-left = 0
padding-right = 2

modules-left = text

tray-position = right
tray-padding = 2
tray-transparent = true
tray-background = ${colors.background}

[module/text]
type = custom/text
content = " "

; vim:ft=dosini
```
If I remove the `tray-transparent` key, I get the following, which looks much better:

**With compton:**
![tray_not_transparent_compton](https://user-images.githubusercontent.com/2452038/34446622-e061638e-ecdc-11e7-9886-a409274ca4a5.png)
**Without compton:**
![tray_not_transparent_no_compton](https://user-images.githubusercontent.com/2452038/34446623-e08505a0-ecdc-11e7-8f74-b12694a25646.png)

So without the `tray-transparent` key and with compton running, things look pretty good on my system.

AndreasBackx commented:
Any update on getting this merged? Would love to use it.

lmoesch commented:
@AndreasBackx Since there is no _real_ interest in merging this, just fork the main branch and merge this request into it yourself - atleast thats what I did.

AichP commented:
@lmoesch Same.

bennofs commented:
@patrick96 @NBonaparte any chance to get this merged? or are there things I need to change before this is ready?

patrick96 commented:
Hi, sorry for the radio silence. This is quite a big PR and I don't have the knowledge to fully understand everything you're doing here, I also don't have the time to do so. I will need to rely on lots of manual testing to make sure it works properly instead of looking for issues in the code. Don't worry we haven't forgotten about this, but there is a lot of things that needs tending right now, so things just kinda get pushed onto the TODO stack.

I cannot promise you that I will look at this in the next week, but it's on my TODO list.

I have one question right away: Polybar already has some transparency functionality, but I know, it's not perfect. What does this PR achieve? Can you please quickly outline what transparency problems that polybar had before are being fixed by this PR

bennofs commented:
Thanks for the quick response! I will try to summarize the changes here:

The transparency functionality that polybar already supports relies on a compositing manager. It sets the background of the window to a (semi-) transparent color and then some compositing manager is needed to compose that with the actual background. 

The problem with that is if you don't use a compositing manager (and many users of minimal window managers don't), then windows with semi-transparent background simply don't work. What you can do however, and this is what the PR here implements, is simply rendering the whole bar directly against the X11 desktop background. So this PR changes the rendering to first render the background (in the pseudo-transparent case, the background is taken from the desktop background) and then rendering all the bar elements above that (using cairo's compositing algorithms to achieve transparency). The window itself does not have a transparent background this way so it also works without a compositing manager. The tradeoff here is that it won't take into account any other windows except the root window's background, so if you move a floating window below the bar then you cannot see that "through the bar" with pseudo-transparency (while with real transparency).

NBonaparte commented:
This also means that the tray background wasn't working properly for us, because it's supposed to work without compositors. Maybe we should try to resolve that issue first.

patrick96 commented:
So that would mean with this patch, we would have pseudo-transparency, even if we had a compositor running, right? If so, how feasible would it be to implement a config option that allows a user to disable pseudo-transparency and use the compositor for transparency?

@NBonaparte from what I found in https://github.com/jaagr/polybar/pull/595#issuecomment-354495390, the tray background works with this patch, even without compton (except for transparent tray icons), as long as you remove the `tray-transparent` key. Or were you referencing the icons themselves not being drawn transparently?

NBonaparte commented:
Yeah, the icon background still isn't transparent if not using a compositor (which I assume would be the most common use case).

patrick96 commented:
Is it even possible to render the icons with an alpha channel, without using a compositor? At least without writing lots of code.

NBonaparte commented:
On second thought, I think this probably might be a case of Qt icons acting funny in the case of Dropbox (see #854), so we likely won't be able to fix it.

patrick96 commented:
I'm not around until the 22nd of july. I will test this once I get back, as soon as I can.

bennofs commented:
Ok. I have now implemented some of the suggested changes:

* I removed `tray-transparent`. The tray background color can now be set to any color with alpha values.
* There is a new option `pseudo-transparency` in `[settings]` that controls whether the bar uses pseudo transparency or real transparency (default is real transparency)
* The tray uses pseudo-transparency in all cases. Changing this would be an invasive change that requires a lot of testing (I don't know if all applications support the `_NET_SYSTEM_TRAY_VISUAL` property). The tray is going to be refactored sometime in the future anyway (#913) so I decided to not spend too much time on this issue.

patrick96 commented:
My exams are starting next week, so I don't have that much time right now. I haven't forgotten about this though.

patrick96 commented:
Just finally tested this and it really looks awesome now, the tray blends in perfectly. Will do some more testing and also do an actual code review.

Banbeucmas commented:
Nice, I am hoping to see implementation for this

tgharib commented:
I forked the repo and merged this PR. Pseudo-transparency in the bar is working but the system tray doesn't seem to be transparent. Is anyone's tray pseudo-transparency working for them? Here is my partial config:

```
[settings]
pseudo-transparency = true

[bar/menu]
monitor = DVI-D-0
monitor-fallback = LVDS1
bottom = true
width = 100%
height = 32

background = #00000000
foreground = #ccffffff

line-color = #00000000
line-size = 2

spacing = 3
padding-right = 2
module-margin-left = 0
module-margin-right = 4

font-0 = NotoSans-Regular:size=8;0
font-1 = unifont:size=6;-3
font-2 = FontAwesome:size=8;-2
font-3 = NotoSans-Regular:size=8;-1
font-4 = MaterialIcons:size=10;-1
font-5 = Termsynu:size=8:antialias=false;0
     
modules-left = i3
modules-right = wireless-network battery volume date
     
tray-position = right
tray-maxsize = 22
tray-background = #00000000
```

bennofs commented:
@tgharib I can reproduce this. Will investigate later, I have some exams this week.
But it looks like this is somehow dependent on whether the bar is bottom or not (it works if you set bottom = false for me)

bennofs commented:
@tgharib can you check with the latest commit applied? I think it should fix your issue

tgharib commented:
It works now! :)

bennofs commented:
Thanks for the feedback! I believe issue 3 is impossible to fix, if my theory about why it occurs is correct:

* i3 resizes the bar window to full screen width
* without a compositing manager, there is no way to set the window bg color to transparent so you will see black

As the window is enlarged by i3, we can't really fix that (we request a smaller window, so to fix it we would need to draw outside of the window).

I'll look into the other issues.

patrick96 commented:
Thought as much for i3, it really doesn't seem to like polybar. Best practice is to use override-redirect with a full-length fully transparent bar below, so this shouldn't be an issue

Banbeucmas commented:
I might want to try this in bspwm and see if the same issue presist like i3

patrick96 commented:
@Banbeucmas I already did, it doesn't happen there.

Banbeucmas commented:
@patrick96 ah ok. Thanks for the info :D

bennofs commented:
@patrick96 the commit I pushed just now should fix both issues. additionally it also reduces the amount of changes, I don't know why I didn't implement it like that in the first place :)

Can you verify that it works? I will make another commit with docs/comments afterwards

patrick96 commented:
Everything seems to check out now :smiley:

bennofs commented:
@patrick96 added comments. should be finished now

patrick96 commented:
I'm sorry, I found another thing:
When the bar displays no text at all and compton isn't running, the bar is always drawn fully opaque.
```dosini
[bar/example]
background = #99000000
foreground = #dfdfdf

modules-left = title

[module/title]
type = internal/xwindow
label =

[settings]
pseudo-transparency = true

; vim:ft=dosini
```

I noticed this because the `xwindow` module displays nothing when no window is focused and I had a bar with only the xwindow module on it.

bennofs commented:
@patrick96 good catch, thanks for testing this so thoroughly! The issue is that `fill_background` uses `compositing-background` as compositing mode which defaults to `source`, thereby overwriting our background image. 

What's the intended purpose of "compositing-background"? Isn't the background supposed to be "the lowest layer", so there's nothing to composite against?

patrick96 commented:
To be honest, I have no idea why it's set to source, all the other settings use `over` and indeed with `over` we don't have that problem. As far as I can tell the change to `source` was made in 3e4a4ce78f3de0081de594940a2a65aede91f990, I don't know why though.

I also don't know what the purpose of `compositing-background` in general is, I think this was before my time. @jaagr can you shed some light on this?

We could probably force `compositing-background = over` if pseudo-transparency is enabled.

bennofs commented:
@patrick96 i believe source makes sense in the non-pseudo-transparent case, since the surface may have undefined contents at that point (?) so compositing might produce unwanted effects. I don't see why one would want to set this to anything other than "SOURCE" though in the non-pseudo-transparent case.

I have pushed a commit that defaults `compositing-background` to `over` if using pseudo-transparency. If somebody really wants to use `source` with pseudo-transparency, they can still do that but polybar will print a warning in that case.

pabloariasal commented:
This feature is fantastic and long overdue! Can't wait to get this merged, @bennofs amazing job! polybar is the only reason I'm forced to use a compositor.

patrick96 commented on src/components/renderer.cpp:
```diff
@@ -160,14 +164,25 @@ renderer::renderer(
   m_comp_border = m_conf.get<cairo_operator_t>("settings", "compositing-border", m_comp_border);
 
+  if (m_comp_bg == CAIRO_OPERATOR_SOURCE && m_pseudo_transparency) {
+    m_log.warn("pseudo-transparency may not work correctly with settings.compositing-background = source");
```
This should check if something other than `over` was used. The warning should also tell them what to do, instead of what not to do.

patrick96 commented:
@NBonaparte can you do a quick sanity check review on this to make sure I didn't miss anything major.

bennofs commented:
Ah, I think it was the compositing-background changes that broke it.

bennofs commented:
@patrick96 I have pushed a new commit that should hopefully fix this. It also means I don't have to change anything about `compositing_background`, so the part about the message is also solved this way.

Please take your time to test this, according to my own tests all examples you gave should work now but as history shows this appears to be tricky to get right. :grimacing:

patrick96 commented on src/components/renderer.cpp:
```diff
@@ -310,6 +300,24 @@ void renderer::end() {
+
+  if (m_pseudo_transparency) {
+    cairo_pattern_t* barcontents{};
+    m_surface->write_png("/tmp/test.png");
```
Is this debugging code or is this actually used?

bennofs replied:
Oops yes obviously the write_PNG is debugging, good catch

patrick96 commented on src/components/renderer.cpp:
```diff
@@ -310,6 +300,24 @@ void renderer::end() {
   }
 
+
+  if (m_pseudo_transparency) {
```
Please document the purpose of this section

bennofs replied:
should be fixed with latest commit

patrick96 commented:
Awesome! I think we're done. I'm going to wait a week on NBonaparte's for feedback but after that we'll merge this.

patrick96 commented:
After only 17 months we have finally merged this!

@bennofs, thank you so much for all your hard work, we really appreciate it. This is a really exciting feature to have!

Would you be available to help out if problems arise with this feature?

bennofs commented:
Thanks! You can ping me in any issues related to this.

patrick96 commented:
Awesome! Let's hope we won't need to ;)

Banbeucmas commented:
@patrick96 @bennofs Finally, congratulations.

kingkobra87 commented:
Thank you so much for this!

fominv commented:
Thank you guys for your work!

pabloariasal commented:
Thank you so much for this!! @bennofs you are amazing!

patrick96 commented:
Can confirm. @bennofs any idea how to fix this? I was able to get transparency back by adding the offset to both the x and y arguments [here](https://github.com/jaagr/polybar/pull/595/files#diff-6b4943a2b2ebf4e9ec7dfc1953c02d0eR406) (`cairo_set_source_surface(*m_context, *surface, -x, -y);`). But that gives me the background of wherever the tray would be without offset. Is it even possible to get the background for areas that are not covered by the bar?

bennofs commented:
I will look into this when I have time (probably right after new year, possibly earlier)

patrick96 commented:
No rush :)

bennofs commented:
> 

> Looks like setting an offset x value for the tray breaks the transparency : [example](https://cdn.discordapp.com/attachments/280379502440480768/526073138619088917/unknown.png)

The example file seems to have gone away. Can you restore it / upload to github?

Phaenyxx commented:
Didn't know then I could just paste a clipped image, here you go.
![image](https://user-images.githubusercontent.com/27176415/50576886-e3c8bc00-0e1b-11e9-8bef-6630587d45a6.png)

bennofs commented:
Ok. What exactly is the intended behaviour of that? The space that is "left over" due to the x/y movement should be filled with the bar's background as well?

patrick96 commented:
I think the intended behavior should be that the whole tray has a transparent background even the parts that do not overlap the bar window.

I think the gap that the tray leaves doesn't need to be filled.

bennofs commented:
Should be fixed by #1571
